### PR TITLE
Make link command use auth crate instead of config token

### PR DESCRIPTION
**Pull request by Zertsov**
### Description
`turbo link` was using the `BaseCommand` config token which exclusively reads from `config.json`, instead of using the new auth crate to find the token. This PR fixes that and uses the auth crate to find the token.

### Testing Instructions
Ensure you don't have a `config.json` with a token in it in whatever `XDG_CONFIG_HOME` you have (where we keep our tokens), run a `turbo login` and/or `turbo login --sso-team <team>`, then run the `turbo link` command. It should properly link.

After that, to double make sure, go to the current version of the code (in `main`) and try the same. If you have the `auth.json` that we create, with a valid token, but no `config.json`, then `turbo link` should fail.

Closes TURBO-1976

### CODE REVIEW

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
Gets us into a better state, but we need to make sure the correct token gets used in `run` and other code paths that require auth.

File: crates/turborepo-lib/src/commands/mod.rs [hunk 1]
```diff
@@ -2,9 +2,9 @@ use std::cell::OnceCell;
 
 use sha2::{Digest, Sha256};
 use turbopath::{AbsoluteSystemPath, AbsoluteSystemPathBuf};
-use turborepo_api_client::{APIAuth, APIClient};
+use turborepo_api_client::{APIAuth, APIClient, Client};
 use turborepo_auth::{
-    TURBOREPO_AUTH_FILE_NAME, TURBOREPO_CONFIG_DIR, TURBOREPO_LEGACY_AUTH_FILE_NAME,
+    AuthFile, TURBOREPO_AUTH_FILE_NAME, TURBOREPO_CONFIG_DIR, TURBOREPO_LEGACY_AUTH_FILE_NAME,
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/commands/mod.rs:7]:
```suggestion
   TURBOREPO_AUTH_FILE_NAME, TURBOREPO_CONFIG_DIR, TURBOREPO_LEGACY_AUTH_FILE_NAME,
```

File: crates/turborepo-lib/src/commands/mod.rs [hunk 2]
```diff
@@ -125,15 +125,27 @@ impl CommandBase {
...
         let team_slug = config.team_slug();
 
-        let Some(token) = config.token() else {
-            return Ok(None);
+        let auth_file_path = self.global_auth_path()?;
+        let config_file_path = self.global_config_path()?;
+        let client = self.api_client()?;
+        let auth = match turborepo_auth::read_auth_file(&auth_file_path) {
+            Ok(auth) => auth,
+            Err(_) => match turborepo_auth::read_auth_file(&config_file_path) {
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/commands/mod.rs:133]: I think this always fails. We're trying to deserialize the old `config.json` as an `auth.json` and it'll be missing the `tokens` field.

File: crates/turborepo-auth/src/lib.rs [hunk 3]
```diff
@@ -89,6 +89,30 @@ pub async fn read_or_create_auth_file(
     Ok(auth_file)
 }
 
+/// Sync version of `read_or_create_auth_file`.
```
Comment (chris-olszewski) [inline crates/turborepo-auth/src/lib.rs:92]: I think the comment needs to be updated as there are behavior differences between this and `read_or_create_auth_file` (and also they're both now sync).

File: crates/turborepo-auth/src/lib.rs [hunk 4]
```diff
@@ -89,6 +89,30 @@ pub async fn read_or_create_auth_file(
...
+            .map_err(|e| Error::FailedToReadAuthFile {
+                source: e,
+                path: auth_file_path.to_owned(),
+            })?;
+        let tokens: AuthFile = serde_json::from_str(&content)
+            .map_err(|e| Error::FailedToDeserializeAuthFile { source: e })?;
+        let mut auth_file = AuthFile::new();
+        for (api, token) in tokens.tokens() {
+            auth_file.insert(api.to_owned(), token.to_owned());
+        }
```
Comment (chris-olszewski) [inline crates/turborepo-auth/src/lib.rs:106]: I don't understand why we're rebuilding the auth file here

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
Primarily concerned with not being able to use the old tokens

File: crates/turborepo-auth/src/lib.rs (hunk 4 above)
Comment (Zertsov) [reply crates/turborepo-auth/src/lib.rs:106]: Lazy copy-paste job - I'll fix that in a moment.

File: crates/turborepo-lib/src/commands/mod.rs (hunk 2 above)
Comment (Zertsov) [reply crates/turborepo-lib/src/commands/mod.rs:133]: Yeah, I can just use the existing read function now that it's sync. I'll clean up.

File: crates/turborepo-auth/src/auth_file.rs [hunk 5]
```diff
@@ -88,10 +87,10 @@ impl AuthToken {
 }
 
 /// Converts our old style of token held in `config.json` into the new schema.
-pub fn convert_to_auth_token(token: &str, client: &impl Client) -> AuthToken {
+pub fn convert_to_auth_token(token: &str, api: &str) -> AuthToken {
```
Comment (chris-olszewski) [inline crates/turborepo-auth/src/auth_file.rs:90]: For a future PR this makes more sense as `AuthToken::new`

**Review by chris-olszewski — APPROVED** [review]

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
Test failures look legit. It looks like we're not respecting `--token` during `run` now.

**Review by chris-olszewski — APPROVED** [review]

**Pull request merged**